packages/driver/src/cypress/cy.ts

```diff
@@ -1257,11 +1259,100 @@ export class $Cy extends EventEmitter2 implements ITimeouts, IStability, IAssert
 
     args.unshift(subject)
 
-    this.Cypress.action('cy:next:subject:prepared', subject, args, firstCall)
-
     return args
   }
 
+  /*
+   * Use `currentSubject()` to get the subject. It reads from cy.state('subjects'), but the format and details of
+   * determining this should be considered an internal implementation detail of Cypress, subject to change at any time.
+   *
+   * Currently, state('subjects') is an object, mapping chainerIds to the current subject for that chainer. For
+   * example, it might look like:
+   *
+   * {
+   *   'chainer2': 'foobar',
+   *   'chainer4': <input>,
+   * }
+   *
+   * Do not read this directly; Prefer currentSubject() instead.
+   */
+  currentSubject (chainerId = this.state('chainerId')) {
+    return (this.state('subjects') || {})[chainerId]
+  }
+
+  /*
+   * Cypress executes commands asynchronously, and those commands can contain other commands - this means that there
+   * are times when an outer chainer might have as its subject the (as of yet unresolved) return value of the inner
+   * chain of commands.
+   *
+   * cy.state('subjectLinks') is where we store that connection. The exact contents should be considered an internal
+   * implementation detail - do not read or alter it directly, but prefer the public interface (linkSubject and
+   * breakSubjectLinksToCurrentChainer).
+   *
+   * In the current implementation, subjectLinks might look like:
+   * {
+   *   'chainer4': 'chainer2',
+   * }
+   *
+   * indicating that when we eventually resolve the subject of chainer4, it should *also* be used as the subject for
+   * chainer2 - for example, `cy.then(() => { return cy.get('foo') }).log()`. The inner chainer (chainer4,
+   * `cy.get('foo')`) is linked to the outer chainer (chainer2) - when we eventually .get('foo'), the resolved value
+   * becomes the new subject for the outer chainer.
+   *
+   * Whenever we are in the middle of resolving one chainer and a new one is created, Cypress links the inner chainer
+   * to the outer one. This is *usually* desireable, allowing simple constructs like
+   * `cy.then(() => { return cy.get('foo') }).log()` to function intuitively.
+   *
+   * But you don't always want to use the inner chainer's subject for the outer chainer. Consider:
+   * `cy.then(() => { cy.get('foo').click(); return 'success' }).log()`
+   *
+   * In this case, we want to break the connection between the inner chainer and the outer one, so that we can
+   * instead use the return value as the new subject. Is this case, you'll want cy.breakSubjectLinksToCurrentChainer().
+   */
+  linkSubject (fromChainerId, toChainerId) {
+    const links = this.state('subjectLinks') || {}
+
+    links[fromChainerId] = toChainerId
```

Comment:
is `fromChainerId` the child / chained command? ex:
cy.get().should().should() or c1 -> c2 -> c3 would map to:
```
 {
   'c2': 'c1',
   'c3': 'c1,
  }
```

Reply:
Yep, exactly. I will add a note to change this to child / parent in my next PR, that will be clearer variable names.

Reply:
ahhh thats right. 👍🏻 thanks!